### PR TITLE
[Windows] Prepare keyboard & text input plugins for multi-view

### PR DESCRIPTION
**Pull request by loic-sharma**
The keyboard & text input plugins are singletons, however, they are currently owned by the view. This change makes them shared by the `FlutterWindowsEngine` to prepare for a multi-view world.

Part of https://github.com/flutter/flutter/issues/115611

## Keyboard

Type | Category | Notes
-- | -- | --
`KeyboardManager` | 1 per window | Receives Windows messages for key events. If the message isn't handled by Flutter, it redispatches the message back to Windows and then ignores Windows's redelivery of the message.<br/><br/>The `KeyboardManager` is owned by the `Window`, but in the future it will be refactored to be owned by the `FlutterWindowsView`. Currently it cannot be 1 per engine as redispatched messages' state must be tracked separately for each `Window`.
`KeyboardKeyHandler` | 1 per engine | Delegates keyboard events to `KeyboardKeyEmbedderHandler` and `KeyboardKeyChannelHandler`. Previously `KeyboardKeyHandler` was 1 per view.
`KeyboardKeyEmbedderHandler` | 1 per engine | Sends Flutter key events from the engine to the framework through the embedder API. Previously `KeyboardKeyEmbedderHandler` was 1 per view.
`KeyboardKeyChannelHandler` | 1 per engine | Sends raw key events from the engine to the framework through a method channel. Previously `KeyboardKeyChannelHandler` was 1 per view.

## Text input

Type | Category | Notes
-- | -- | --
`TextInputPlugin` | 1 per engine | Manages text input. Previously `TextInputPlugin` was 1 per view.
`TextInputPluginDelegate` | 1 per view | Used by `TextInputPlugin` to update the window as the user interacts with text input. For example, it updates the window's caret.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: shell/platform/windows/keyboard_unittests.cc
Comment (loic-sharma) [inline shell/platform/windows/keyboard_unittests.cc:341]: This helper sends a platform message and isn't actually related to the view in any way. I moved it to the `KeyboardTester` and renamed it `InjectPlatformMessage`.

**Comment by dkwingsmt** [comment]
The choice of making the embedder handler and the channel handler per-engine kind of confuses me. Because I'm hoping to merge the embedder handler into the keyboard manager once the channel handler is removed. They are separate only because one event needs to "branch" into two sets of logic right now. In this way we'll only have one class (at least in the perspective of the engine) to handle all keyboard logic. I'm guess you did this because they're connected to global engine API (sending and receiving events). If so, shall we extract these API out as a delegate?

**Comment by loic-sharma** [comment]
Thanks for the feedback!

> maybe we could extract these API out as a delegate?

Could you expand on this?

> I'm hoping to merge the embedder handler into the keyboard manager ...

From my understanding, we will need to track Windows keyboard messages on a per-view basis for redispatching purposes. That said, I completely agree that a bunch of logic in `KeyboardManager` should be moved out. However, I don't think that change is necessary before making the keyboard multi-view ready. I'm concerned this change may break keyboard/text input in unexpected ways, so I'd like to minimize risk as much as possible.

> .. now that the keyboard manager is per-view

I haven't changed the keyboard manager yet, it is still attached to the underlying `Window`. I state above that there is 1 keyboard manager per view since there is 1 view per window :)

**Review by yaakovschectman — APPROVED** [review]
I'm also a bit unclear how you decided what would be per-view or per-engine. LGTM aside from that and existing comments.

**Comment by loic-sharma** [comment]
> It seems like this won't affect anything in the framework like keyboard listeners, focus, text input connection, etc?

Yup, that's the goal.

> I'm also a bit unclear how you decided what would be per-view or per-engine

I updated the description to try and make that more clear. TLDR: Everything that could be 1 per engine today is now 1 per engine. :

1. `TextInputPluginDelegate` - It interacts with the underlying `Window` on behalf of the `TextInputPlugin`. For example, it'll update the window's caret.
2. `KeyboardManager` - Amongst other things, this tracks state for each `Window`/`FlutterWindowsView`. We do plan to refactor this to move logic up to the engine, however, that will be done in a subsequent change.

**Comment by dkwingsmt** [comment]
> Everything that could be 1 per engine today is now 1 per engine. 

Are you saying the two keyboard handler classes should stay 1-per-engine? Well, I'm down going either path of making the keyboard system 1-per-engine or 1-per-view, but then we should make all of keyboard manager and the two handler classes the same.

**Comment by dkwingsmt** [comment]
After offline discussion, it seems that I had some misunderstanding. The current plan is that eventually we'll move as much of the keyboard logic as possible into engine, except for the part that needs to be view-specific, i.e. the redispatching handling. I agree with this plan.

**Comment by chinmaygarde** [comment]
From Triage: This looks good to go. May I land it?

**Comment by loic-sharma** [comment]
@chinmaygarde I'd like to get approval from @dkwingsmt before landing this. Is there something that's blocked by this?

**Review by dkwingsmt — APPROVED** [review]
LGTM

**Pull request merged**
